### PR TITLE
Make ITransport top-level abstraction for sending events to Sentry.

### PR DESCRIPTION
**Pull request by maciejwalkowiak**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring

## :scroll: Description

Flip the relation between `AsyncConnection` and `ITransport` and make `BlockingHttpTransport` use `HttpConnection`, making it possible to provide alternative transport implementations not bound to use thread pool.

## :bulb: Motivation and Context

In order to provide non blocking transport, transport itself has to be responsible for managing threads (not the AsyncConnection).

## :green_heart: How did you test it?

Unit & integration tests.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [ ] No breaking changes

### CODE REVIEW

**Comment by maciejwalkowiak** [comment]
@marandaneto, as a next step - I believe we should create an interface `ITransportFactory` and put it on `SentryOptions`, instead of `ITransfer` itself. Currently, since transport is set on SentryOptions, but transport itself also likely depends on SentryOptions, it may be difficult for users to provide custom transport implementations.

File: sentry/src/main/java/io/sentry/transport/BlockingHttpTransport.java [hunk 1]
```diff
@@ -11,57 +11,79 @@
...
+    }
+
+    final SentryEnvelope filteredEnvelope = rateLimiter.filter(envelope, hint);
+
+    if (filteredEnvelope == null) {
+      if (cached) {
+        envelopeCache.discard(envelope);
+      }
+    } else {
+      executor.submit(new EnvelopeSender(filteredEnvelope, hint, currentEnvelopeCache));
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/transport/BlockingHttpTransport.java:85]: I think the name `BlockingHttpTransport` is misleading since it gives the impression the interface is going to block until for the event submission, which isnt' the case.

We're scheduling the event through the executor here.

Maybe we can use another name?

File: sentry/src/main/java/io/sentry/transport/StdoutTransport.java [hunk 2]
```diff
@@ -15,14 +15,13 @@ public StdoutTransport(final @NotNull ISerializer serializer) {
...
-  public TransportResult send(final @NotNull SentryEnvelope envelope) throws IOException {
+  public void send(final @NotNull SentryEnvelope envelope, Object hint) throws IOException {
     Objects.requireNonNull(envelope, "SentryEnvelope is required");
 
     try {
       serializer.serialize(envelope, System.out);
-      return TransportResult.success();
     } catch (Exception e) {
-      return TransportResult.error();
+      // do nothing
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/transport/StdoutTransport.java:24]: not really against but why the change?

File: sentry/src/main/java/io/sentry/transport/BlockingHttpTransport.java (hunk 1 above)
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/transport/BlockingHttpTransport.java:85]: yeah I kinda agree, not sure about a suitable name though.

SingleThreadHttpTransport?

File: sentry/src/main/java/io/sentry/transport/HttpConnection.java [hunk 3]
```diff
@@ -27,37 +24,21 @@
...
 
-/**
- * An implementation of the {@link ITransport} interface that sends the events to the Sentry server
- * over HTTP(S).
- */
 @Open
-@ApiStatus.NonExtendable // only not final because of testing
-@ApiStatus.Internal
-public class HttpTransport implements ITransport {
+class HttpConnection {
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/transport/HttpConnection.java:32]: is the `HttpConnection` reusable for the NIO implementation as well? just trying to understand why we have the BlockingHttpTransport + HttpConnection, I do get their difference, but what would be the advantage of it right now?

**Comment by marandaneto** [comment]
do we need a changelog? how do we communicate the breaking changes later on?

File: sentry/src/main/java/io/sentry/transport/BlockingHttpTransport.java (hunk 1 above)
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/transport/BlockingHttpTransport.java:85]: Can be also `AsyncHttpTransport`. It is async, but blocking.

File: sentry/src/main/java/io/sentry/transport/StdoutTransport.java (hunk 2 above)
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/transport/StdoutTransport.java:24]: The interface has changed and it does not return anymore.

File: sentry/src/main/java/io/sentry/transport/HttpConnection.java (hunk 3 above)
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/transport/HttpConnection.java:32]: Just to split the responsibilities - and I agree that the split is not perfect. I tried to create decorators like `EnvelopeCachingHttpTransport` and `RateLimitingHttpTransport` but since `send` returns `void` it's also not possible. 

`HttpConnection` is package private, to give us flexibility of refactoring later without breaking users' code.

**Comment by maciejwalkowiak** [comment]
@marandaneto @bruno-garcia I need this PR reviewed.

File: sentry/src/main/java/io/sentry/AsyncHttpTransportFactory.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import io.sentry.util.Objects;
+import java.net.MalformedURLException;
+import java.net.URL;
+import org.jetbrains.annotations.NotNull;
+
+/** Creates {@link AsyncHttpTransport}. */
+final class AsyncHttpTransportFactory implements TransportFactory {
+
+  @Override
+  public ITransport create(final @NotNull SentryOptions options) {
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/AsyncHttpTransportFactory.java:16]: should we add `throws IllegalArgumentException` in the method signature?

Comment (marandaneto) [inline sentry/src/main/java/io/sentry/AsyncHttpTransportFactory.java:16]:
```suggestion
  public @NotNull ITransport create(final @NotNull SentryOptions options) {
```

File: sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java [hunk 5]
```diff
@@ -139,7 +122,6 @@ public void close() throws IOException {
                 "Failed to shutdown the async connection async sender within 1 minute. Trying to force it now.");
         executor.shutdownNow();
       }
-      transport.close();
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java:142]: how does the connection get closed now?

File: sentry/src/main/java/io/sentry/transport/HttpConnection.java (hunk 3 above)
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/transport/HttpConnection.java:32]: sounds good to me

File: sentry/src/main/java/io/sentry/transport/HttpConnection.java [hunk 6]
```diff
@@ -27,37 +24,21 @@
...
-  private final @Nullable SSLSocketFactory sslSocketFactory;
-  private final @Nullable HostnameVerifier hostnameVerifier;
-
-  private final @NotNull SentryOptions options;
-
-  private final @NotNull ILogger logger;
-
-  private final @NotNull RateLimiter rateLimiter;
+  private final SentryOptions options;
+  private final RateLimiter rateLimiter;
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/transport/HttpConnection.java:41]:
```suggestion
  private final @NotNull SentryOptions options;
  private final @NotNull RateLimiter rateLimiter;
```

File: sentry/src/main/java/io/sentry/transport/HttpConnection.java [hunk 7]
```diff
@@ -67,51 +48,31 @@
...
+  public HttpConnection(
       final @NotNull SentryOptions options,
       final @NotNull IConnectionConfigurator connectionConfigurator,
-      final int connectionTimeoutMillis,
-      final int readTimeoutMillis,
-      final @Nullable SSLSocketFactory sslSocketFactory,
-      final @Nullable HostnameVerifier hostnameVerifier,
-      final @NotNull URL sentryUrl) {
+      final @NotNull URL sentryUrl,
+      RateLimiter rateLimiter) {
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/transport/HttpConnection.java:58]:
```suggestion
      final @NotNull RateLimiter rateLimiter) {
```

File: sentry/src/main/java/io/sentry/transport/HttpConnection.java [hunk 8]
```diff
@@ -67,51 +48,31 @@
...
       final @NotNull SentryOptions options,
       final @NotNull IConnectionConfigurator connectionConfigurator,
-      final int connectionTimeoutMillis,
-      final int readTimeoutMillis,
-      final @Nullable SSLSocketFactory sslSocketFactory,
-      final @Nullable HostnameVerifier hostnameVerifier,
       final @NotNull URL sentryUrl,
       final @NotNull AuthenticatorWrapper authenticatorWrapper,
-      final @NotNull RateLimiter rateLimiter) {
+      RateLimiter rateLimiter) {
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/transport/HttpConnection.java:72]:
```suggestion
      final @NotNull RateLimiter rateLimiter) {
```

File: sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java (hunk 5 above)
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java:142]: well the method was not implemented anyway I guess

**Review by marandaneto — APPROVED** [review]
few missing jetbrains annotations but other than that LGTM.

I didn't find a plugin that fails the build if annotations are missing, but it'd be nice to write a custom linter for that, so I don't need to be the human linter here :D

I am only picky about that cus it's nicer when writing Kotlin code.

File: sentry/src/main/java/io/sentry/AsyncHttpTransportFactory.java (hunk 4 above)
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/AsyncHttpTransportFactory.java:16]: I think we should not, what's the reasonable way to handle this exception here anyways?

Comment (marandaneto) [reply sentry/src/main/java/io/sentry/AsyncHttpTransportFactory.java:16]: if we handle it down its fine, I was just curious cus the `TransportFactory` now is the way to go for custom transports right

File: sentry/src/main/java/io/sentry/TransportFactory.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package io.sentry;
+
+import io.sentry.transport.ITransport;
+import org.jetbrains.annotations.NotNull;
+
+/** Creates instances of {@link ITransport}. */
+public interface TransportFactory {
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/TransportFactory.java:7]: should we do `ITransportFactory`?

Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/TransportFactory.java:7]: I have mixed feelings about putting `I` in the front of names for interfaces, but I will just follow your preference. Tell me :-)

Comment (marandaneto) [reply sentry/src/main/java/io/sentry/TransportFactory.java:7]: no feelings or preferences here, but see `ITransport`, so ideally, we do a single way, either all of them or none of them.
chose which ones to rename :)

Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/TransportFactory.java]: Let it be `ITransportFactory`.

File: sentry/src/main/java/io/sentry/AsyncHttpTransportFactory.java (hunk 4 above)
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/AsyncHttpTransportFactory.java]: We don't handle it, the exception is propagated to the very top and in the end Sentry init will fail.

This PR is a PR to another branch, so once whole package is there we can take a look at it again.
